crates/key-value-aws/src/store.rs

```diff
@@ -0,0 +1,528 @@
+    Unknown,
+}
+
+struct CompareAndSwap {
+    key: String,
+    client: Client,
+    table: Arc<String>,
+    bucket_rep: u32,
+    state: Mutex<CasState>,
+}
+
+/// Primary key in DynamoDB items used for querying items
+const PK: &str = "PK";
+/// Value key in DynamoDB items storing item value as binary
+const VAL: &str = "VAL";
+/// Version key in DynamoDB items used for atomic operations
+const VER: &str = "VER";
+
+#[async_trait]
+impl Store for AwsDynamoStore {
+    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>, Error> {
+        let response = self
+            .client
+            .get_item()
+            .consistent_read(self.consistent_read)
+            .table_name(self.table.as_str())
+            .key(
+                PK,
+                aws_sdk_dynamodb::types::AttributeValue::S(key.to_string()),
+            )
+            .projection_expression(VAL)
+            .send()
+            .await
+            .map_err(log_error)?;
+
+        let item = response.item.and_then(|mut item| {
+            if let Some(AttributeValue::B(val)) = item.remove(VAL) {
+                Some(val.into_inner())
+            } else {
+                None
+            }
+        });
+
+        Ok(item)
+    }
+
+    async fn set(&self, key: &str, value: &[u8]) -> Result<(), Error> {
+        self.client
+            .put_item()
+            .table_name(self.table.as_str())
+            .item(PK, AttributeValue::S(key.to_string()))
+            .item(VAL, AttributeValue::B(Blob::new(value)))
+            .send()
+            .await
+            .map_err(log_error)?;
+        Ok(())
+    }
+
+    async fn delete(&self, key: &str) -> Result<(), Error> {
+        if self.exists(key).await? {
+            self.client
+                .delete_item()
+                .table_name(self.table.as_str())
+                .key(PK, AttributeValue::S(key.to_string()))
+                .send()
+                .await
+                .map_err(log_error)?;
+        }
+        Ok(())
+    }
+
+    async fn exists(&self, key: &str) -> Result<bool, Error> {
+        let GetItemOutput { item, .. } = self
+            .client
+            .get_item()
+            .consistent_read(self.consistent_read)
+            .table_name(self.table.as_str())
+            .key(
+                PK,
+                aws_sdk_dynamodb::types::AttributeValue::S(key.to_string()),
+            )
+            .projection_expression(PK)
+            .send()
+            .await
+            .map_err(log_error)?;
+
+        Ok(item.map(|item| item.contains_key(PK)).unwrap_or(false))
+    }
+
+    async fn get_keys(&self) -> Result<Vec<String>, Error> {
+        self.get_keys().await
+    }
+
+    async fn get_many(&self, keys: Vec<String>) -> Result<Vec<(String, Option<Vec<u8>>)>, Error> {
+        let mut results = Vec::with_capacity(keys.len());
+
+        if keys.is_empty() {
+            return Ok(results);
+        }
```

Comment:
Nit: `with_capacity` does allocation which we can avoid in the empty key case by moving the initialization down.

```suggestion
        if keys.is_empty() {
            return Ok(Vec::new());
        }
        
         let mut results = Vec::with_capacity(keys.len());
```

Reply:
As a side note, I see [potentially unnecessary vec allocations at that level](https://github.com/ogghead/spin/blob/main/crates/factor-key-value/src/host.rs#L287)

Reply:
I made the changes at the higher level (and removed vec copying there) -- hopefully this makes sense, let me know if it's preferable to leave this empty list check at the AWS level